lib/SyntaxParse/SyntaxTreeCreator.cpp

```diff
@@ -109,32 +116,29 @@ SyntaxTreeCreator::realizeSyntaxRoot(OpaqueSyntaxNode rootN,
   return rootNode;
 }
 
-OpaqueSyntaxNode
-SyntaxTreeCreator::recordToken(tok tokenKind,
-                               ArrayRef<ParsedTriviaPiece> leadingTriviaPieces,
-                               ArrayRef<ParsedTriviaPiece> trailingTriviaPieces,
-                               CharSourceRange range) {
-  size_t leadingTriviaLen =
-    ParsedTriviaPiece::getTotalLength(leadingTriviaPieces);
-  size_t trailingTriviaLen =
-    ParsedTriviaPiece::getTotalLength(trailingTriviaPieces);
-  SourceLoc tokLoc = range.getStart().getAdvancedLoc(leadingTriviaLen);
-  unsigned tokLength = range.getByteLength() -
-      leadingTriviaLen - trailingTriviaLen;
-  CharSourceRange tokRange = CharSourceRange{tokLoc, tokLength};
-  SourceLoc leadingTriviaLoc = range.getStart();
-  SourceLoc trailingTriviaLoc = tokLoc.getAdvancedLoc(tokLength);
-  Trivia syntaxLeadingTrivia =
-    ParsedTriviaPiece::convertToSyntaxTrivia(leadingTriviaPieces,
-                                             leadingTriviaLoc, SM, BufferID);
-  Trivia syntaxTrailingTrivia =
-    ParsedTriviaPiece::convertToSyntaxTrivia(trailingTriviaPieces,
-                                             trailingTriviaLoc, SM, BufferID);
-  StringRef tokenText = SM.extractText(tokRange, BufferID);
+OpaqueSyntaxNode SyntaxTreeCreator::recordToken(tok tokenKind,
+                                                StringRef leadingTrivia,
+                                                StringRef trailingTrivia,
+                                                CharSourceRange range) {
+  unsigned tokLength =
+      range.getByteLength() - leadingTrivia.size() - trailingTrivia.size();
+  auto leadingTriviaStartOffset =
+      SM.getLocOffsetInBuffer(range.getStart(), BufferID);
+  auto tokStartOffset = leadingTriviaStartOffset + leadingTrivia.size();
+  auto trailingTriviaStartOffset = tokStartOffset + tokLength;
+
+  // Get StringRefs of the token's texts that point into the syntax arena's
+  // buffer.
+  StringRef leadingTriviaText =
+      ArenaSourceBuffer.substr(leadingTriviaStartOffset, leadingTrivia.size());
+  StringRef tokenText = ArenaSourceBuffer.substr(tokStartOffset, tokLength);
+  StringRef trailingTriviaText = ArenaSourceBuffer.substr(
+      trailingTriviaStartOffset, trailingTrivia.size());
+
   auto ownedText = OwnedString::makeRefCounted(tokenText);
```

Comment:
Now that `RawSyntax` always have `SyntaxArena` and `SyntaxArena` has `ArenaSourceBuffer`. Do we really need to make ref-counted owned string? I feel storing `tokenText` in `RaySyntax` is sufficient.

Reply:
Maybe we can just abolish `OwnedString` and `copyToArenaIfNeeded(tokenText)` just like trivia.

Reply:
That's my next PR ;-) I wanted to not inflate this even further.

Reply:
Great!